packages/powersync_attachments_stream/lib/src/sync/syncing_service.dart

```diff
@@ -0,0 +1,324 @@
+class SyncingService {
+  final AbstractRemoteStorageAdapter remoteStorage;
+  final AbstractLocalStorageAdapter localStorage;
+  final AbstractAttachmentService attachmentsService;
+  final Future<String> Function(String) getLocalUri;
+  final AbstractSyncErrorHandler? errorHandler;
+  final Duration syncThrottle;
+  final Duration period;
+  final Logger logger;
+
+  StreamSubscription? _syncSubscription;
+  StreamSubscription? _periodicSubscription;
+  bool _isClosed = false;
+  final _syncTriggerController = StreamController<void>.broadcast();
+
+  SyncingService({
+    required this.remoteStorage,
+    required this.localStorage,
+    required this.attachmentsService,
+    required this.getLocalUri,
+    this.errorHandler,
+    this.syncThrottle = const Duration(seconds: 5),
+    this.period = const Duration(seconds: 30),
+    Logger? logger,
+  }) : logger = logger ?? Logger('SyncingService');
+
+  /// Starts the syncing process, including periodic and event-driven sync operations.
+  ///
+  /// [period] is the interval at which periodic sync operations are triggered.
+  Future<void> startSync({Duration period = const Duration(seconds: 30)}) async {
+    if (_isClosed) return;
+
+    _syncSubscription?.cancel();
+    _periodicSubscription?.cancel();
+
+    // Create a merged stream of manual triggers and attachment changes
+    final attachmentChanges = attachmentsService.watchActiveAttachments();
+    final manualTriggers = _syncTriggerController.stream;
+
+    // Merge both streams and apply throttling
+    final mergedStream = StreamGroup.merge([attachmentChanges, manualTriggers])
+        .transform(_throttleTransformer(syncThrottle))
+        .listen((_) async {
+          try {
+            await attachmentsService.withContext((context) async {
+              final attachments = await context.getActiveAttachments();
+              logger.info(
+                'Found ${attachments.length} active attachments',
+              );
+              await handleSync(attachments, context);
+              await deleteArchivedAttachments(context);
+            });
+          } catch (e, st) {
+            if (e is! StateError && e.toString().contains('cancelled')) {
+              logger.severe(
+                'Caught exception when processing attachments',
+                e,
+                st,
+              );
+            } else {
+              rethrow;
+            }
+          }
+        });
+
+    _syncSubscription = mergedStream;
+
+    // Start periodic sync
+    _periodicSubscription = Stream.periodic(period).listen((_) {
+      logger.info('Periodically syncing attachments');
+      triggerSync();
+    });
+  }
+
+  StreamTransformer<T, T> _throttleTransformer<T>(Duration throttle) {
+    return StreamTransformer.fromHandlers(
+      handleData: (data, sink) {
+        sink.add(data);
+        // Simple throttle implementation - just delay the next event
+        Future.delayed(throttle);
```

Comment:
This future is unawaited and will not actually throttle anything.

I think it would be easier to not throttle the stream here and instead add a duration parameter to `watchActiveAttachments`. We can forward that to `db.watch()` instead of duplicating the throttle implementation here. That means we wouldn't be throttling `manualTriggers` anymore, but I think that's fine.